File: js/leetcode/A173_Binary_Search_Tree_Iterator.js
/**
Implement an iterator over a binary search tree (BST). Your iterator will be initialized with the root node of a BST.

Calling next() will return the next smallest number in the BST.

Note: next() and hasNext() should run in average O(1) time and uses O(h) memory, where h is the height of the tree.
 */

class BSTIterator {
    constructor(root) {
        this.stack = [];
        let p = root;
        while (p) {
            this.stack.push(p);
            p = p.left;
        }
    }
    hasNext() {
        return this.stack.length > 0;
    }
    next() {
        let curr = this.stack.pop();
        let p = curr.right;
        while (p) {
            this.stack.push(p);
            p = p.left;
        }
        return curr.val;
    }
}

File: java/src/leetcode/problems/A173_Binary_Search_Tree_Iterator.java
package leetcode.problems;

import leetcode.datastructures.TreeNode;

import java.util.*;

class BSTIterator {
    Stack<TreeNode> stack;

    public BSTIterator(TreeNode root) {
        stack = new Stack<>();
        pushAll(root);
    }

    public boolean hasNext() {
        return !stack.isEmpty();
    }

    public int next() {
        TreeNode curr = stack.pop();
        pushAll(curr.right);
        return curr.val;
    }

    private void pushAll(TreeNode node) {
        while (node != null) {
            stack.push(node);
            node = node.left;
        }
    }
}
